File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    
    
    public ListNode reverse(ListNode start,ListNode end){
        ListNode front=start,last=null,ptr=start;
        while(front!=end){
            front=front.next;
            ptr.next=last;
            last=ptr;
            ptr=front;
        }
        return last;
    }
    
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode node=head;
        for(int i=0;i<k;i++){
            if(node==null)return head;
            node=node.next;
        }
        ListNode Nnode=reverse(head,node);
        head.next=reverseKGroup(node,k);
        return Nnode;
        
    }
}

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    
    // ListNode*recur(ListNode*start,ListNode*end){
    //     ListNode*front=start,*ptr=start,*back=NULL;
    //     while(front!=end){
    //         front=front->next;
    //         ptr->next=back;
    //         back=ptr;
    //         ptr=front;
    //     }
    //     return back;
    // }
    
    ListNode*reverse(ListNode*st,ListNode*end){
        ListNode*forw=st,*curr=st,*prev=NULL;
            while(forw!=end){
                forw=forw->next;
                curr->next=prev;
                prev=curr;
                curr=forw;
                
            }
            // curr->next=prev;
            return prev;
    }
    
    
    ListNode* reverseKGroup(ListNode* head, int k) {
        // ListNode*temp=head;
        // for(int i=0;i<k;i++){
        //     if(!temp){
        //         return head;
        //     }
        //     temp=temp->next;
        // }
        // ListNode*Nnode=recur(head,temp);
        // head->next=reverseKGroup(temp,k);
        // return Nnode;
        
        ListNode*list=head;
       
        for(int i=0;i<k;i++){
            if(list)
                list=list->next;
            else
                return head;
        }
        ListNode*nNode= reverse(head,list);
        head->next=reverseKGroup(list,k);
        return nNode;        
        
    }
};